src/cutlass_mlp.cu

```diff
@@ -314,6 +327,606 @@ void CutlassMLP<T>::backward_impl(
+			bool return_hidden_dL2dw = compute_dL2dw<FullLayer, T>(
+				multi_streams.back().get(1), // stream
+				false,
+				m_activation,
+				weight_matrix_at(use_inference_params, i),
+				forward.hidden.at(tmp_idx-1), // input
+				fc_output.at(tmp_idx), // p
+				dL1dp.at(m_n_hidden_matmuls - bwd_idx - 1), // dL1dp
+				dL1doutput.at(m_n_hidden_matmuls - bwd_idx - 1), // dL1doutput
+				dL2_ddL1doutput.at(bwd_bwd_idx-1), // dL2d_dL1dinput
+				ddoutputdp_dp.at(i+1), // ddoutputdp_dp
+				dL2_ddL1doutput.at(bwd_bwd_idx), // dL2_ddL1doutput
+				gradient_matrix_at(i), // gradient matrix
+				param_gradients_mode
+			);
+
+			//multi_streams.emplace_back(stream, 2);
+			bool return_hidden_dL2dinput = compute_dL2dinput<FullLayer, T>(
+				stream, //multi_streams.back().get(3), // stream,
+				false,
+				m_activation,
+				weight_matrix_at(use_inference_params, i), // weights
+				fc_output.at(i+1), // p
+				dL1doutput.at(m_n_hidden_matmuls - i - 1), // dL_doutput
+				dL2_ddL1doutput.at(bwd_bwd_idx-1), // dL2_ddL1dinput of current layer
+				ddoutputdp_dp.at(i+1), // ddoutputdp_dp
+				dL2dinput.at(i+1), // dL2dinput of current layer
+				param_gradients_mode
+			);
+
+			bool return_hidden_dL2_ddL1doutput = compute_dL2_ddL1doutput<FullLayer, T>(
+				stream, //multi_streams.back().get(2), //stream,
+				false, // is_inference
+				m_activation,
+				weight_matrix_at(use_inference_params, i),
+				fc_output.at(tmp_idx), // p,
+				dL2_ddL1doutput.at(bwd_bwd_idx-1), // dL2_ddL1dinput
+				dL2_ddL1doutput.at(bwd_bwd_idx), // dL2_ddL1doutput
+				param_gradients_mode
+			);
+
+			tmp_idx++; // tmp_idx += can_fused ? 1 : 2;
+			bwd_bwd_idx++;
+			bwd_idx++;
+		}
+
+		// Output layer weight
+		multi_streams.emplace_back(stream, 2);
+		bool return_output_dL2dw = compute_dL2dw<FullLayer, T>(
+			multi_streams.back().get(1), //stream, 
+			false,
+			m_output_activation,
+			output_weight_matrix(use_inference_params),
+			forward.hidden.at(tmp_idx-1), // input
+			fc_last_output, // p
+			m_output_activation == Activation::None ? dL_doutput : backward_output_tmp, // dL1dp
+			dL_doutput, // dL1doutput
+			dL2_ddL1doutput.at(bwd_bwd_idx-1), // dL2d_dL1dinput
+			ddoutputdp_dp.at(m_n_hidden_matmuls+1), // ddoutputdp_dp
+			*dL_ddLdoutput, // dL2_ddL1doutput
+			output_gradient_matrix(),
+			param_gradients_mode
+		);
+
+		// Output layer dL2dinput
+		bool return_output_dL2dinput = compute_dL2dinput<FullLayer, T>(
+			stream,
+			false, // is_inference
+			m_output_activation,
+			output_weight_matrix(use_inference_params), // weights
+			fc_last_output, // p
+			m_output_activation == Activation::None ? dL_doutput : backward_output_tmp, // dL_doutput
+			dL2_ddL1doutput.at(bwd_bwd_idx-1), // dL2_ddL1dinput of current layer
+			ddoutputdp_dp.at(m_n_hidden_matmuls+1), // ddoutputdp_dp
+			dL2dinput.at(m_n_hidden_matmuls+1), // dL2dinput of current layer
+			param_gradients_mode
+		);		
+
+		if (dL_ddLdoutput) { // if dL_ddLdoutput is not nullptr
```

Comment:
Redundant comments like this one should be removed from the PR. (Basically, use comments only to explain behavior that is *not* clear from the programming language itself. Like algorithmic details, mathematical reasoning, hidden details, etc.)